### PR TITLE
Compatibility with carta-casacore updated to CASA 6.6.0

### PR DESCRIPTION
**Pull request by pford**
**Description**

* What is implemented or fixed? Mention the linked issue(s), if any.
Changes for compatibility with carta-casacore, updated to use casa6 release 6.6.0 and newer casacore.

* How does this PR solve the issue? Give a brief summary.
Fixes for casacore changes: (1) can no longer pass std::vector implicitly for casacore::Vector parameter, (2) casacore function to determine data type has changed signature.  The casacore update removes std::binary_function from casacore IPosition.h (removed in C++17).

* Are there any companion PRs (frontend, protobuf)?
CARTAvis/carta-casacore#25

* Is there anything else that testers should know (e.g. exactly how to reproduce the issue)?
Build backend with updated carta-casacore and all tests pass.

**Checklist**

- [x] changelog updated / ~no changelog update needed~
- [x] e2e test passing / ~corresponding fix added~ / ~new e2e test created~
- [x] ~protobuf updated to the latest dev commit~ / no protobuf update needed
- [x] ~protobuf version bumped~ / protobuf version not bumped
- [ ] added reviewers and assignee
- [ ] added ZenHub estimate, milestone, and release

### CODE REVIEW

**Comment by kswang1029** [comment]
@pford @ajm-asiaa after rebuilding/reinstalling https://github.com/CARTAvis/carta-casacore/pull/25 I can build the backend successfully. However, when I try to load any image, the backend crashes immediately. The backend log did not say much except a "segmentation fault". 

My build env: 
- macOS: Sonoma 14.0
- clang: 15.0.0
- cmake: 3.22.1

My cmake log:

> -- The C compiler identification is AppleClang 15.0.0.15000040
-- The CXX compiler identification is AppleClang 15.0.0.15000040
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /Library/Developer/CommandLineTools/usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /Library/Developer/CommandLineTools/usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found OpenMP_C: -Xclang -fopenmp (found version "5.0") 
-- Found OpenMP_CXX: -Xclang -fopenmp (found version "5.0") 
-- Found OpenMP: TRUE (found version "5.0")  
-- Found PkgConfig: /opt/homebrew/bin/pkg-config (found version "0.29.2") 
-- Found gsl using find_package
-- Found ZFP: /usr/local/lib/cmake/zfp/zfp-config.cmake (found version "0.5.5") 
-- Checking for one of the modules 'cfitsio'
-- Found cfitsio using pkg-config
-- Checking for one of the modules 'wcslib'
-- Found wcslib using pkg-config
-- Checking for one of the modules 'uuid'
-- Found uuid using pkg-config
-- Checking for one of the modules 'libzstd'
-- zstd found using pkg-config
-- Found HDF5: /opt/homebrew/Cellar/hdf5/1.12.1/lib/libhdf5_cpp.dylib;/opt/homebrew/Cellar/hdf5/1.12.1/lib/libhdf5.dylib;/opt/homebrew/opt/szip/lib/libsz.dylib;/Library/Developer/CommandLineTools/SDKs/MacOSX14.0.sdk/usr/lib/libz.tbd;/Library/Developer/CommandLineTools/SDKs/MacOSX14.0.sdk/usr/lib/libdl.tbd;/Library/Developer/CommandLineTools/SDKs/MacOSX14.0.sdk/usr/lib/libm.tbd (found version "1.12.1") found components: CXX 
-- Found Protobuf: /opt/homebrew/lib/libprotobuf.dylib (found version "3.19.1") 
-- Found Threads: TRUE  
-- Build spdlog: 1.8.2
-- Build type: RelWithDebInfo
-- Configuring done
-- Generating done
-- Build files have been written to: /Users/kswang/carta_build/carta-backend/build


The macOS crash log is attached below just in case it is helpful.

> Thread 0 Crashed::  Dispatch queue: com.apple.main-thread
0   libcasa_casa.7.dylib          	       0x102b6efb0 casacore::LogSink::~LogSink() + 72
1   libcasa_tables.7.dylib        	       0x1037ae72c casacore::LoggerHolderRep::LoggerHolderRep(bool) + 180
2   libcasa_tables.7.dylib        	       0x1037adb14 casacore::LoggerHolder::LoggerHolder(bool) + 44
3   libcasa_images.7.dylib        	       0x102e6dc88 casacore::ImageInterface<float>::ImageInterface() + 72
4   libcasa_images.7.dylib        	       0x102e336b4 casacore::FITSImage::FITSImage(casacore::String const&, unsigned int, unsigned int) + 44
5   carta_backend                 	       0x100c5f828 carta::FitsLoader::AllocateImage(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&) + 1088 (FitsLoader.cc:94)
6   carta_backend                 	       0x100c447a8 carta::FileLoader::OpenFile(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&) + 36 (FileLoader.cc:93)
7   carta_backend                 	       0x100bb80d8 carta::FileExtInfoLoader::FillFileInfoFromImage(CARTA::FileInfoExtended&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>&) + 72 (FileExtInfoLoader.cc:132)
8   carta_backend                 	       0x100bb7ab0 carta::FileExtInfoLoader::FillFileExtInfo(CARTA::FileInfoExtended&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>&) + 408 (FileExtInfoLoader.cc:102)
9   carta_backend                 	       0x100bb20cc carta::FileExtInfoLoader::FillFitsFileInfoMap(std::__1::map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, CARTA::FileInfoExtended, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const, CARTA::FileInfoExtended>>>&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>&) + 440 (FileExtInfoLoader.cc:74)
10  carta_backend                 	       0x100da1fd4 carta::Session::FillExtendedFileInfo(std::__1::map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, CARTA::FileInfoExtended, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const, CARTA::FileInfoExtended>>>&, CARTA::FileInfo&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>&) + 376 (Session.cc:254)
11  carta_backend                 	       0x100da6590 carta::Session::OnFileInfoRequest(CARTA::FileInfoRequest const&, unsigned int) + 160 (Session.cc:468)
12  carta_backend                 	       0x100dc91ac carta::SessionManager::OnMessage(uWS::WebSocket<false, true, carta::PerSocketData>*, std::__1::basic_string_view<char, std::__1::char_traits<char>>, uWS::OpCode) + 3016 (SessionManager.cc:250)
13  carta_backend                 	       0x100dcf904 ofats::any_detail::any_invocable_impl<void, false, uWS::WebSocket<false, true, carta::PerSocketData>*, std::__1::basic_string_view<char, std::__1::char_traits<char>>, uWS::OpCode>::call(uWS::WebSocket<false, true, carta::PerSocketData>*, std::__1::basic_string_view<char, std::__1::char_traits<char>>, uWS::OpCode) + 24 (MoveOnlyFunction.h:247) [inlined]
14  carta_backend                 	       0x100dcf904 ofats::any_invocable<void (uWS::WebSocket<false, true, carta::PerSocketData>*, std::__1::basic_string_view<char, std::__1::char_traits<char>>, uWS::OpCode)>::operator()(uWS::WebSocket<false, true, carta::PerSocketData>*, std::__1::basic_string_view<char, std::__1::char_traits<char>>, uWS::OpCode) + 24 (MoveOnlyFunction.h:354) [inlined]
15  carta_backend                 	       0x100dcf904 uWS::WebSocketContext<false, true, carta::PerSocketData>::handleFragment(char*, unsigned long, unsigned int, int, bool, uWS::WebSocketState<true>*, void*) + 1240 (WebSocketContext.h:100)
16  carta_backend                 	       0x100dce6ac bool uWS::WebSocketProtocol<true, uWS::WebSocketContext<false, true, carta::PerSocketData>>::consumeMessage<6u, unsigned char>(unsigned char, char*&, unsigned int&, uWS::WebSocketState<true>*, void*) + 756 (WebSocketProtocol.h:363)
17  carta_backend                 	       0x100dce274 uWS::WebSocketProtocol<true, uWS::WebSocketContext<false, true, carta::PerSocketData>>::consume(char*, unsigned int, uWS::WebSocketState<true>*, void*) + 176 (WebSocketProtocol.h:476)
18  carta_backend                 	       0x100dce160 auto uWS::WebSocketContext<false, true, carta::PerSocketData>::init()::'lambda'(auto*, char*, int)::operator()<us_socket_t>(auto*, char*, int) const + 120 (WebSocketContext.h:306)
19  carta_backend                 	       0x100dce0dc auto uWS::WebSocketContext<false, true, carta::PerSocketData>::init()::'lambda'(auto*, char*, int)::__invoke<us_socket_t>(auto*, char*, int) + 32 (WebSocketContext.h:284)
20  carta_backend                 	       0x100e91474 us_loop_run + 204 (epoll_kqueue.c:147)
21  carta_backend                 	       0x100dcc4e8 uWS::Loop::run() + 4 (Loop.h:182) [inlined]
22  carta_backend                 	       0x100dcc4e8 uWS::run() + 12 (Loop.h:199) [inlined]
23  carta_backend                 	       0x100dcc4e8 uWS::TemplatedApp<false>::run() + 12 (App.h:571) [inlined]
24  carta_backend                 	       0x100dcc4e8 carta::SessionManager::RunApp() + 276 (SessionManager.cc:593)
25  carta_backend                 	       0x100d18df0 main + 3588 (Main.cc:220)
26  dyld                          	       0x187509058 start + 2224

**Comment by ajm-ska** [comment]
Unfortunately, I confirm the same "Segmentation fault: 11" when opening an image. My crash log looks the same.

- macOS: Ventura 13.0.1 Intel
- clang: 14.0.0
- cmake: 3.25.2

**Comment by ajm-ska** [comment]
@kswang1029 I think I see what happened. If you go into `carta-casacore/casa6` and do `git log` does it show `6.5.0.4`?

I need to do `git checkout 842f3d` in casa6 in order to change it to `6.6.0.20`.

I'm not sure why it didn't change after doing `git checkout pam/update_casa6` and `git submodule update --init --recursive`.

**Comment by pford** [comment]
I also cd to carta-casacore/casa6 and `git submodule update` for casacore, which should be at a43b76.

I am also seeing the seg fault.  I think my branch worked at the time I opened the PR, but when I updated it with dev (including the casacore log fix) it broke.  I am looking into it.

**Comment by pford** [comment]
@kswang1029 @ajm-asiaa I added a fix, please try again and see if it works now.

**Comment by kswang1029** [comment]
@pford @ajm-asiaa with the new commits, now I can load images successfully. Thanks for the fix. I will try to run the full e2e test suite to see if there are unexpected issues.

**Comment by ajm-ska** [comment]
@pford Thanks! It appears to work fine for me now on Intel macOS Ventura 13 with Chrome 117, Firefox 118 and Safari 16.1.

Please ignore the failing CI for now. We still need to update the CI platforms with pam/update_casa6 carta-casacore, and then ensure this carta-backend branch is merged with all existing branches so that the CI will not fail on the other branches.

**Review by kswang1029 — APPROVED** [review]
In addition to macOS Ventura, I also tested with ubuntu 22.04 and it works. On both platforms, no regression is observed via e2e tests. 👍

**Comment by kswang1029** [comment]
> In addition to macOS Ventura, I also tested with ubuntu 22.04 and it works. On both platforms, no regression is observed via e2e tests. 👍

Will need to redo the test to ensure correct casa6 commit is used.

**Review by kswang1029 — APPROVED** [review]
After building carta-casacore in the right way (https://github.com/CARTAvis/carta-casacore/pull/25#pullrequestreview-1692883664), I see no regression from e2e tests on macOS Sonoma and ubuntu 22.04.

**Review by ajm-ska — APPROVED** [review]
Maybe we can announce at the group meeting on Thursday that this PR is ready to be merged? Then I can update carta-casacore on the CI platforms so that it will pass the tests, and developers can update their branches with it. What do you think @kswang1029?

**Comment by kswang1029** [comment]
> Maybe we can announce at the group meeting on Thursday that this PR is ready to be merged? Then I can update carta-casacore on the CI platforms so that it will pass the tests, and developers can update their branches with it. What do you think @kswang1029?

👌

**Review by confluence — APPROVED** [review]

File: src/ImageData/CartaFitsImage.cc [hunk 1]
```diff
@@ -475,7 +475,8 @@ void CartaFitsImage::SetFitsHeaderStrings(int nheaders, const std::string& heade
     }
 
     // For setting up image
-    _image_header_strings = no_history_strings;
+    casacore::Vector<casacore::String> const no_history_stringsV(no_history_strings);
```
Comment (veggiesaurus) [inline src/ImageData/CartaFitsImage.cc]: `no_history_stringsV` doesn't fit with our variable naming scheme

File: src/Region/RegionHandler.cc [hunk 2]
```diff
@@ -1274,10 +1274,11 @@ bool RegionHandler::CalculatePvPreviewImage(int frame_id, int preview_id, bool q
             // Progress for loading data here if needed due to prior cancel
             if (preview_cube->GetRegionProfile(bounding_box, box_mask, progress_callback, profile, max_num_pixels, message)) {
                 // spdlog::debug("PV preview profile {} of {} max num pixels={}", iregion, num_regions, max_num_pixels);
+                casacore::Vector<float> const profileV(profile);
```
Comment (veggiesaurus) [inline src/Region/RegionHandler.cc]: same here

**Review by veggiesaurus — CHANGES_REQUESTED** [review]

**Pull request merged**